Spigot-API-Patches/0300-PlayerKickEvent-causes.patch

```diff
@@ -0,0 +1,133 @@
+index 5c0efe74237dbe6803ce023fde99682ff70d1a92..e416fe46549b96f14db37885e0db13d2b28a931f 100644
+--- a/src/main/java/org/bukkit/event/player/PlayerKickEvent.java
++++ b/src/main/java/org/bukkit/event/player/PlayerKickEvent.java
+@@ -12,6 +12,7 @@ public class PlayerKickEvent extends PlayerEvent implements Cancellable {
+     private static final HandlerList handlers = new HandlerList();
+     private net.kyori.adventure.text.Component leaveMessage; // Paper
+     private net.kyori.adventure.text.Component kickReason; // Paper
++    private final Cause cause; // Paper
+     private Boolean cancel;
+ 
+     @Deprecated // Paper
+@@ -19,14 +20,25 @@ public class PlayerKickEvent extends PlayerEvent implements Cancellable {
+         super(playerKicked);
+         this.kickReason = org.bukkit.Bukkit.getUnsafe().legacyComponentSerializer().deserialize(kickReason); // Paper
+         this.leaveMessage = org.bukkit.Bukkit.getUnsafe().legacyComponentSerializer().deserialize(leaveMessage); // Paper
++        this.cause  = Cause.UNKNOWN; // Paper
+         this.cancel = false;
+     }
+     // Paper start
++    @Deprecated
+     public PlayerKickEvent(@NotNull final Player playerKicked, @NotNull final net.kyori.adventure.text.Component kickReason, @NotNull final net.kyori.adventure.text.Component leaveMessage) {
+         super(playerKicked);
+         this.kickReason = kickReason;
+         this.leaveMessage = leaveMessage;
+         this.cancel = false;
++        this.cause = Cause.UNKNOWN;
++    }
++
++    public PlayerKickEvent(@NotNull final Player playerKicked, @NotNull final net.kyori.adventure.text.Component kickReason, @NotNull final net.kyori.adventure.text.Component leaveMessage, @NotNull final Cause cause) {
++        super(playerKicked);
++        this.kickReason = kickReason;
++        this.leaveMessage = leaveMessage;
++        this.cancel = false;
++        this.cause = cause;
+     }
+ 
+     /**
+@@ -132,4 +144,69 @@ public class PlayerKickEvent extends PlayerEvent implements Cancellable {
+     public static HandlerList getHandlerList() {
+         return handlers;
+     }
++    // Paper start
++    /**
++     * Gets the cause of this kick
++     *
++     * @return
++     */
++    @NotNull
++    public org.bukkit.event.player.PlayerKickEvent.Cause getCause() {
++        return cause;
++    }
++
++    public enum Cause {
++
++        PLUGIN,
++
++        WHITELIST,
++
++        BANNED,
++
++        IP_BANNED,
++
++        KICK_COMMAND,
++
++        FLYING_PLAYER,
++
++        FLYING_VEHICLE,
++
++        TIMEOUT,
++
++        IDLING,
++
++        INVALID_VEHICLE_MOVEMENT,
++
++        INVALID_PLAYER_MOVEMENT,
++
++        INVALID_ENTITY_ATTACKED,
++
++        INVALID_PAYLOAD,
++
++        SPAM,
++
++        INVALID_HOTBAR_SELECTION,
++
++        DROP_ITEMS_TOO_QUICK,
++
++        BOOK_TOO_LARGE,
++
++        BOOK_EDIT_TOO_QUICK,
```

Comment:
I dislike these; they're mitigations for various other issues which may be dropped in the future. Perhaps instead just have a `CHEATING`, `ILLEGAL_ACTION`, or something else?

Reply:
K, bundled them up into `ILLEGAL_ACTION`, they all fit that moniker